### PR TITLE
Loads

### PR DESCRIPTION
**Pull request by tyler-french**
Add the "+" Prefix loading method

### CODE REVIEW

File: repo/repo.go [hunk 1]
```diff
@@ -98,16 +98,46 @@ func ListRepositories(workspace *rule.File) (repos []*rule.Rule, repoFileMap map
...
 			macroFile, err := rule.LoadMacroFile(f, "", defName)
 			if err != nil {
 				return nil, nil, err
 			}
 			for _, repo := range macroFile.Rules {
-				if name := repo.Name(); name != "" {
-					repos = append(repos, repo)
-					repoFileMap[name] = macroFile
-					repoIndexMap[name] = len(repos) - 1
+				if kind := repo.Kind(); kind == "go_repository" || !leveled {
```
Comment (tyler-french) [inline repo/repo.go:109]: Using the OR not leveled - I avoid needing any additional logic

File: repo/repo.go [hunk 2]
```diff
@@ -98,16 +98,46 @@ func ListRepositories(workspace *rule.File) (repos []*rule.Rule, repoFileMap map
...
+					}
+					fi, err := rule.LoadMacroFile(callFile, "", kind)
+					if err != nil {
+						return nil, nil, err
+					}
+					for _, repo := range fi.Rules {
+						if name := repo.Name(); name != "" {
+							repos = append(repos, repo)
+							repoFileMap[name] = fi
+							repoIndexMap[name] = len(repos) - 1
```
Comment (tyler-french) [inline repo/repo.go:137]: repeated 3 times -> could be separate func

File: repo/repo.go (hunk 1 above)
Comment (linzhp) [reply repo/repo.go:109]: Maybe:
```go
if repo.Kind() ==  "go_repository" {
//...
} else if leveled {
//...
}

Comment (linzhp) [reply repo/repo.go:109]: We can still support other kinds of repos when `leveled`:
```go
if name := repo.Name(); name != "" {
//...
} else if leveled {
// treat all function calls without `name` as macros
}
```

File: repo/repo.go [hunk 3]
```diff
@@ -98,16 +98,45 @@ func ListRepositories(workspace *rule.File) (repos []*rule.Rule, repoFileMap map
...
+				if kind := repo.Kind(); kind == "go_repository" || !leveled {
+					if name := repo.Name(); name != "" {
+						repos = append(repos, repo)
+						repoFileMap[name] = macroFile
+						repoIndexMap[name] = len(repos) - 1
+					}
+				} else {
+					// This block will only be entered if leveled==true
+					var callFile string
+					var defName string
```
Comment (linzhp) [inline repo/repo.go:118]: nit: `var callFile, defName string`

File: repo/repo.go [hunk 4]
```diff
@@ -98,16 +98,45 @@ func ListRepositories(workspace *rule.File) (repos []*rule.Rule, repoFileMap map
 			if err != nil {
 				return nil, nil, err
 			}
+			leveled := strings.HasPrefix(f, "+")
+			f = strings.TrimPrefix(f, "+")
 			f = filepath.Join(filepath.Dir(workspace.Path), filepath.Clean(f))
 			macroFile, err := rule.LoadMacroFile(f, "", defName)
```
Comment (linzhp) [inline repo/repo.go:104]: Extract a  function `loadRepositoriesFromMacro` from here, so we can  recursively  call it

File: repo/repo.go [hunk 5]
```diff
@@ -98,16 +98,45 @@ func ListRepositories(workspace *rule.File) (repos []*rule.Rule, repoFileMap map
...
+						if l.Has(kind) {
+							callFile = filepath.Join(filepath.Dir(workspace.Path), filepath.Clean(l.Name()))
+							defName = l.Get(kind)
+							break
+						}
+					}
+					if len(callFile) == 0 {
+						continue
+					}
+					fi, err := rule.LoadMacroFile(callFile, "", defName)
```
Comment (linzhp) [inline repo/repo.go:129]: recursively call `loadRepositoriesFromMacro` from here

File: repo/repo.go [hunk 6]
```diff
@@ -98,16 +98,45 @@ func ListRepositories(workspace *rule.File) (repos []*rule.Rule, repoFileMap map
...
+					var defName string
+					for _, l := range macroFile.Loads {
+						if l.Has(kind) {
+							callFile = filepath.Join(filepath.Dir(workspace.Path), filepath.Clean(l.Name()))
+							defName = l.Get(kind)
+							break
+						}
+					}
+					if len(callFile) == 0 {
+						continue
```
Comment (linzhp) [inline repo/repo.go:127]: Add a `TODO` if  you don't want to handle in this PR the situation where one macro calls another macro in the same bzl file, hence no matching `load` statement

File: rule/rule.go [hunk 7]
```diff
@@ -592,6 +592,11 @@ func (l *Load) Has(sym string) bool {
 	return ok
 }
 
+// Has returns the original (from) name of a (to) Symbol from a load.
+func (l *Load) Get(sym string) string {
```
Comment (linzhp) [inline rule/rule.go:596]: `Get` is too ambiguous. Maybe `Unalias`?

**Pull request closed without merging**